xla/service/gpu/gpu_hlo_cost_analysis.cc

```diff
@@ -316,12 +336,65 @@ int64_t FlopsPerElement(const absl::string_view device_name,
   return FindOrDefault(dtype_profiles->get(), opcode, kDefaultFlopsPerElement);
 }
 
+int64_t GetFlopsForElementwiseOp(const GpuDeviceInfo* gpu_device_info,
+                                 const HloOpcode op_code, const Shape& shape) {
+  int64_t flop_per_element = FlopsPerElement(
+      gpu_device_info ? gpu_device_info->name : kDefaultDeviceName,
+      shape.element_type(), op_code);
+  return flop_per_element * ShapeUtil::ElementsInRecursive(shape);
+}
+
+int64_t GetFlopsForElementwiseOp(const GpuDeviceInfo* gpu_device_info,
+                                 const HloInstruction* instr) {
+  LOG(ERROR) << "#####ShapeUtil::ElementsInRecursive(instr->shape()) "
+             << ShapeUtil::ElementsInRecursive(instr->shape())
+             << " INSTR: " << instr->ToString();
+
+  return GetFlopsForElementwiseOp(gpu_device_info, instr->opcode(),
+                                  instr->shape());
+}
+
+Status GpuHloCostAnalysis::HandleAllReduce(const HloInstruction* allreduce) {
+  const HloModuleConfig& config = allreduce->GetModule()->config();
+  int64_t num_devices = config.num_partitions();
+  int64_t num_replicas = config.replica_count();
+  // Data parallel ranks
+  int64_t num_ranks = num_devices;
```

Comment:
We should look at the number of devices in the replica groups, as in many cases the all-reduce may not be across all devices. all-reduce supports non-uniform replica group size, so we can just use the largest size here. Also note subtleties around collectiveOpGroupMode etc. May be we can add a utility function to collective_ops_utils.cc to find the number if participants given replica groups and other info (similar to GetParticipatingDevices but just number of as we don't have device assignment here).

Reply:
That's a good idea. I'm looking into adding this utility function.

Reply:
I added a helper function in collective_ops_utils to get the count of ranks in each replica group.